File: solutions/typescript/crypto-square/1/crypto-square.ts
export class Crypto {
  private normalizedText: string;
  private rows: number;
  private cols: number;
  private rectangle: string[];

  constructor(plainText: string) {
    // Normalize: remove spaces and punctuation, convert to lowercase
    this.normalizedText = plainText.toLowerCase().replace(/[^a-z0-9]/g, '');
    
    // Calculate rectangle dimensions
    const len = this.normalizedText.length;
    this.cols = Math.ceil(Math.sqrt(len));
    this.rows = Math.ceil(len / this.cols);
    
    // Adjust columns to ensure c >= r and c - r <= 1
    while (this.cols < this.rows || this.cols - this.rows > 1) {
      this.cols++;
      this.rows = Math.ceil(len / this.cols);
    }
    
    // Create the rectangle by splitting into rows
    this.rectangle = [];
    for (let i = 0; i < this.rows; i++) {
      const start = i * this.cols;
      const end = start + this.cols;
      let row = this.normalizedText.slice(start, end);
      // Pad the last row if necessary
      if (row.length < this.cols) {
        row = row.padEnd(this.cols, ' ');
      }
      this.rectangle.push(row);
    }
  }

  get ciphertext(): string {
    if (this.normalizedText.length === 0) {
      return '';
    }
    
    // Read down the columns
    const columns: string[] = [];
    for (let c = 0; c < this.cols; c++) {
      let column = '';
      for (let r = 0; r < this.rows; r++) {
        column += this.rectangle[r][c];
      }
      columns.push(column);
    }
    
    // Join with spaces
    return columns.join(' ');
  }
}

File: solutions/java/crypto-square/1/src/main/java/CryptoSquare.java
class CryptoSquare {
    private final String normalized;

    CryptoSquare(String plaintext) {
        // Normalize: lowercase and remove non-alphanumeric characters
        this.normalized = plaintext.toLowerCase().replaceAll("[^a-z0-9]", "");
    }

    String getCiphertext() {
        if (normalized.isEmpty()) {
            return "";
        }

        int length = normalized.length();
        
        // Calculate rectangle dimensions (c >= r, c - r <= 1, r * c >= length)
        int c = (int) Math.ceil(Math.sqrt(length));
        int r = (int) Math.ceil((double) length / c);
        
        // Pad the normalized string with spaces to fill the rectangle perfectly
        StringBuilder padded = new StringBuilder(normalized);
        while (padded.length() < r * c) {
            padded.append(' ');
        }
        
        // Read down the columns to form chunks
        StringBuilder result = new StringBuilder();
        for (int col = 0; col < c; col++) {
            if (col > 0) {
                result.append(' ');
            }
            for (int row = 0; row < r; row++) {
                result.append(padded.charAt(row * c + col));
            }
        }
        
        return result.toString();
    }
}